---
title: "Analysis of single cell ferret pancreatic ductal differentiation"
author: "Kristen Wells"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    highlight: "tango"
    df_print: "paged"
    code_folding: "hide"
    self_contained: true
---

# Incorrect annotation {.tabset}
Another issue we've uncovered is that some of the 3' annotations seem to be incorrect. For example, you can see that many of the Pdx1 reads are mapping outside of the annotated gene and are not being counted. I have found a tool to bioinformatically predict other regions that may have this problem, but all of them will need to be manually checked.

```{r "plot-annotation", echo = F}
library(here)
library(tidyverse)
library(knitr)
samples <- c("WT_D2", "WT_D5", "WT_D7", "WT_D9", "WT_D14",
             "CFKO_D2", "CFKO_D5", "CFKO_D7", "CFKO_D9", "CFKO_D14")

annotation_template <- file.path(here(),
                         "src/scripts/templates/annotation_template.Rmd")


annotation_chunks <- samples %>%
  map(~knit_expand(file = annotation_template, sample = .x))
```

`r knit_child(text = annotation_chunks)`